xcell/cls/cov.py

```diff
@@ -132,44 +134,54 @@ def get_outdir(self):
         outdir = os.path.join(root, 'cov')
         return outdir
 
-    def get_covariance_workspace(self):
+    def get_covariance_workspace(self, save_cw=True):
         """
         Return the covariance workspace needed to compute the Gaussian
         covariance block. If 'recompute' is not set in the configuration file,
         it will read it from the output directory if found.
 
+        Parameters
+        ----------
+        save_cw: bool
+            If True, save the covariance workspace
+
         Return
         ------
         cw: pymaster.NmtCovarianceWorkspace
             Covariance workspace to compute the Gaussian covariance block
 
         """
+        if self.cw is not None:
+            return self.cw
+
         mask1, mask2 = self.clA1A2.get_masks_names()
         mask3, mask4 = self.clB1B2.get_masks_names()
         fname = os.path.join(self.outdir,
                              f'cw__{mask1}__{mask2}__{mask3}__{mask4}.fits')
         cw = nmt.NmtCovarianceWorkspace()
         recompute = self.data.data['recompute']['cmcm']
-        if recompute or (not os.path.isfile(fname)):
-            n_iter = self.data.data['sphere']['n_iter_cmcm']
-            l_toeplitz, l_exact, dl_band = self.data.check_toeplitz('cov')
-            fA1, fB1 = self.clA1B1.get_nmt_fields()
-            fA2, fB2 = self.clA2B2.get_nmt_fields()
-            cw.compute_coupling_coefficients(fA1, fA2, fB1, fB2,
-                                             n_iter=n_iter,
-                                             l_toeplitz=l_toeplitz,
-                                             l_exact=l_exact,
-                                             dl_band=dl_band)
-            # Recheck again in case other process has started writing it
-            if (not os.path.isfile(fname)):
-                cw.write_to(fname)
-            self.recompute_cmcm = False
-        else:
-            cw.read_from(fname)
-
-        return cw
-
-    def _get_cl_for_cov(self, clab, clab_fid, ma, mb):
+        if (not recompute) and os.path.isfile(fname):
+            tools.read_wsp(cw, fname)
+            if cw.wsp is not None:
+                return cw
+
+        n_iter = self.data.data['sphere']['n_iter_cmcm']
+        l_toeplitz, l_exact, dl_band = self.data.check_toeplitz('cov')
+        fA1, fB1 = self.clA1B1.get_nmt_fields()
+        fA2, fB2 = self.clA2B2.get_nmt_fields()
+        cw.compute_coupling_coefficients(fA1, fA2, fB1, fB2,
+                                         n_iter=n_iter,
+                                         l_toeplitz=l_toeplitz,
+                                         l_exact=l_exact,
+                                         dl_band=dl_band)
+        if save_cw:
+            tools.save_wsp(cw, fname)
+        self.recompute_cmcm = False
+        self.cw = cw
+
+        return self.cw
+
+    def _get_cl_for_cov(self, clab, clab_fid):
```

Comment:
same question, was this just a reshuffling of the above?

Reply:
(beyond the ma, mb thing)

Reply:
Same as for the workspace. Now we use the `read_wsp` functions that know how to handle the situation of loading a corrupted file.